src/watch.ts

```diff
@@ -11,6 +16,126 @@ const enum State {
   Connected,
 }
 
+const enum QueueState {
+  Idle,
+  ReadingRevision,
+  Attaching,
+  Destroyed,
+}
+
+class AttachQueue {
+  private state = QueueState.Idle;
+  private queue: Watcher[] = [];
+  private revision: number;
+
+  constructor(
+    private readonly stream: RPC.IDuplexStream<RPC.IWatchRequest, RPC.IWatchResponse>,
+    private readonly kv: RPC.KVClient,
+  ) {}
+
+  /**
+   * Inserts a watcher to be attached to the stream.
+   */
+  public attach(watcher: Watcher | Watcher[]) {
+    this.queue = this.queue.concat(watcher);
+
+    if (this.state === QueueState.Idle) {
+      this.readQueue();
+    }
+  }
+
+  /**
+   * Remove the watcher from any pending attach queue.
+   */
+  public dequeue(watcher: Watcher) {
+    this.queue = this.queue.filter(w => w !== watcher);
+  }
+
+  /**
+   * Dispatches the "create" response to the waiting watcher and fires the
+   * next one as necessary.
+   */
+  public handleCreate(res: RPC.IWatchResponse) {
+    const watcher = this.queue.shift();
+    if (!watcher) {
+      throw new ClientRuntimeError('Could not find watcher corresponding to create response');
+    }
+
+    (<{ id: string }>watcher).id = res.watch_id;
+    watcher.emit('connected', res);
+    this.readQueue();
+  }
+
+  /**
+   * Halts future operations on the queue.
+   */
+  public destroy() {
+    this.state = QueueState.Destroyed;
+  }
+
+  /**
+   * Reads the next watcher to create off the queue and attaches it.
+   */
+  private readQueue() {
+    if (this.state === QueueState.Destroyed) {
+      return;
+    }
+    if (this.queue.length === 0) {
+      this.state = QueueState.Idle;
+      return;
+    }
+
+    const watcher = this.queue[0];
+    if (watcher.request.start_revision) {
+      if (!this.revision) {
+        return this.readRevision(watcher);
+      }
+
+      watcher.request.start_revision = Math.min(
```

Comment:
I'm sorry, but I did not understand this logic.

If I understood correctly, then:

`this.revision` - revision of the key, the last initialized watcher.
`watcher.request.start_revision` - is a revision on which watcher received an error.

And if you follow the logic, then you should start the watcher strictly from the moment the watcher stopped, i.e. with `watcher.request.revision`.
Or did I not understand something?